src/main/java/org/opensearch/agent/tools/SearchAnomalyDetectorsTool.java

```diff
@@ -123,24 +131,92 @@ public <T> void run(Map<String, String> parameters, ActionListener<T> listener)
+                }
+                CompletableFuture<List<GetAnomalyDetectorResponse>> listFuture = CompletableFuture
+                    .allOf(profileFutures.toArray(new CompletableFuture<?>[0]))
+                    .thenApply(v -> profileFutures.stream().map(CompletableFuture::join).collect(Collectors.toList()));
+                final List<GetAnomalyDetectorResponse> profileResponses = listFuture.join();
+                for (GetAnomalyDetectorResponse profileResponse : profileResponses) {
+                    if (profileResponse != null && profileResponse.getDetector() != null) {
+                        String detectorId = profileResponse.getDetector().getId();
+
+                        // We follow the existing logic as the frontend to determine overall detector state
+                        // https://github.com/opensearch-project/anomaly-detection-dashboards-plugin/blob/main/server/routes/utils/adHelpers.ts#L437
+                        String detectorState;
+                        ADTask realtimeTask = profileResponse.getRealtimeAdTask();
+
+                        if (realtimeTask != null) {
+                            String taskState = realtimeTask.getState();
+                            if (taskState.equalsIgnoreCase("CREATED")) {
+                                detectorState = DetectorStateString.Initializing.name();
+                            } else if (taskState.equalsIgnoreCase("RUNNING")) {
+                                detectorState = DetectorStateString.Running.name();
+                            } else if (taskState.equalsIgnoreCase("INIT_FAILURE")
+                                || taskState.equalsIgnoreCase("UNEXPECTED_FAILURE")
+                                || taskState.equalsIgnoreCase("FAILED")) {
+                                detectorState = DetectorStateString.Failed.name();
+                            } else {
+                                // Task states may fall under other values, such as "FEATURE_REQUIRED" / "STOPPED" / etc.
+                                // Defaulting all other states to fall under the disabled category
+                                detectorState = DetectorStateString.Disabled.name();
+                            }
+                        } else {
+                            detectorState = DetectorStateString.Disabled.name();
+                        }
+
+                        if (running != null && !running && detectorState.equals(DetectorStateString.Running.name())) {
+                            hitsAsMap.remove(detectorId);
+                        }
+                        if (disabled != null && !disabled && detectorState.equals(DetectorStateString.Disabled.name())) {
+                            hitsAsMap.remove(detectorId);
+                        }
+                        if (failed != null && !failed && detectorState.equals(DetectorStateString.Failed.name())) {
+                            hitsAsMap.remove(detectorId);
+                        }
```

Comment:
It seems like we're doing a similar check and action for different detector states? Could we reduces code duplication and centralizes the filtering logic here?

Reply:
The logic itself can't be simplified since they have to be grouped as they currently are, but I will combine into one conditional.

Reply:
maybe something like this:
```suggestion
      private void filterDetectorsByState(Map<String, SearchHit> hitsAsMap, 
                                                   Boolean stateCondition, 
                                                   String targetState, 
                                                   String actualState) {
    if (stateCondition != null && !stateCondition && actualState.equals(targetState)) {
        hitsAsMap.remove(detectorId);
    }
}

filterDetectorsByState(hitsAsMap, running, DetectorStateString.Running.name(), detectorState);
filterDetectorsByState(hitsAsMap, disabled, DetectorStateString.Disabled.name(), detectorState);
filterDetectorsByState(hitsAsMap, failed, DetectorStateString.Failed.name(), detectorState);
```

Reply:
I like this! Thanks for the suggestion, I think this is the most readable and flexible.

Reply:
Could probably do the map removal as part of the conditional above which sets `detectorState`.

Reply:
Given @dbwiddis suggestion already removed one of the conditionals, I'm going to just combine into a single conditional to simplify things instead of adding a helper fn in this case. This number of states shouldn't change or scale out over time.